{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import ast\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the data and convert the secondary structure to a list of tuples\n",
    "data = pd.read_csv('/Users/dominiquefastus/master_project/Data/NEW_ECOLI_FULL_uniprot_02_sec_struc_filtered_final.csv', converters={'secondary_structure': ast.literal_eval}, nrows=1000)\n",
    "data['secondary_structure'] = data['secondary_structure'].apply(lambda x: list(x.items())) \n",
    "\n",
    "# create new columns for the protein sequence split into 3 parts (start, between, end)\n",
    "# create new columns for the nucleotide sequence split into 3 parts (start, between, end)\n",
    "data['protein_sequence_start'], data['protein_sequence_between'], data['protein_sequence_end'] = data['protein_sequence'].str[:20], data['protein_sequence'].str[20:-20], data['protein_sequence'].str[-20:]\n",
    "data['nucleotide_sequence_start'], data['nucleotide_sequence_between'], data['nucleotide_sequence_end'] = data['nucleotide_sequence'].str[:60], data['nucleotide_sequence'].str[60:-60], data['nucleotide_sequence'].str[-60:]\n",
    "\n",
    "# create new columns for the secondary structure split into 3 parts (start, between, end)\n",
    "# convert the secondary structure dictionary to a string\n",
    "data['secstru_sequence_start'], data['secstru_sequence_between'], data['secstru_sequence_end'] = data['secondary_structure'].apply(lambda x: x[:20]), data['secondary_structure'].apply(lambda x: x[20:-20]), data['secondary_structure'].apply(lambda x: x[-20:])\n",
    "data['secstru_sequence_start'], data['secstru_sequence_between'], data['secstru_sequence_end'] = data['secstru_sequence_start'].apply(lambda x: \"\".join(dict(x).values())), data['secstru_sequence_between'].apply(lambda x: \"\".join(dict(x).values())), data['secstru_sequence_end'].apply(lambda x: \"\".join(dict(x).values()))\n",
    "data['secondary_structure'] = data['secondary_structure'].apply(lambda x: \"\".join(dict(x).values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define codon table for amino acids\n",
    "codon_table = {\n",
    "    'F': ['TTT', 'TTC'], 'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "    'I': ['ATT', 'ATC', 'ATA'], 'M': ['ATG'], 'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "    'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'], 'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "    'T': ['ACT', 'ACC', 'ACA', 'ACG'], 'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "    'Y': ['TAT', 'TAC'], 'H': ['CAT', 'CAC'], 'Q': ['CAA', 'CAG'],\n",
    "    'N': ['AAT', 'AAC'], 'K': ['AAA', 'AAG'], 'D': ['GAT', 'GAC'],\n",
    "    'E': ['GAA', 'GAG'], 'C': ['TGT', 'TGC'], 'W': ['TGG'], 'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "    'G': ['GGT', 'GGC', 'GGA', 'GGG'], '*': ['TAA', 'TGA', 'TAG']\n",
    "}\n",
    "\n",
    "def calculate_rscu(nucleotide_sequence):\n",
    "    codon_counts = Counter([nucleotide_sequence[i:i+3] for i in range(0, len(nucleotide_sequence), 3)])\n",
    "    amino_acid_counts = {aa: sum([codon_counts[codon] for codon in codons]) for aa, codons in codon_table.items()}\n",
    "\n",
    "    rscu = {}\n",
    "    for aa, codons in codon_table.items():\n",
    "        total = amino_acid_counts[aa]\n",
    "        for codon in codons:\n",
    "            if total > 0:\n",
    "                rscu[codon] = codon_counts[codon] / (total / len(codons))\n",
    "            else:\n",
    "                rscu[codon] = 0\n",
    "    return rscu\n",
    "\n",
    "# Apply the RSCU calculation to each nucleotide sequence\n",
    "data['rscu_start'] = data['nucleotide_sequence_start'].apply(calculate_rscu)\n",
    "data['rscu_between'] = data['nucleotide_sequence_between'].apply(calculate_rscu)\n",
    "data['rscu_end'] = data['nucleotide_sequence_end'].apply(calculate_rscu)\n",
    "data[['primary_id', 'rscu_start', 'rscu_end']].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to quantify secondary structure motifs\n",
    "def quantify_secondary_structure(structure_sequence):\n",
    "    counts = Counter(structure_sequence)\n",
    "    total = len(structure_sequence)\n",
    "    return {key: counts[key] / total for key in counts}\n",
    "\n",
    "# Apply the function to the secondary structure column\n",
    "data['secondary_structure_quantified_start'] = data['secstru_sequence_start'].apply(quantify_secondary_structure)\n",
    "data['secondary_structure_quantified_between'] = data['secstru_sequence_between'].apply(quantify_secondary_structure)\n",
    "data['secondary_structure_quantified_end'] = data['secstru_sequence_end'].apply(quantify_secondary_structure)\n",
    "data[['primary_id', 'secondary_structure_quantified_start']].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to aggregate RSCU values (e.g., using mean RSCU per protein)\n",
    "def aggregate_rscu(rscu):\n",
    "    return np.mean(list(rscu.values()))\n",
    "\n",
    "# Apply the aggregation function to the RSCU column\n",
    "data['rscu_aggregate_start'] = data['rscu_start'].apply(aggregate_rscu)\n",
    "data['rscu_aggregate_between'] = data['rscu_between'].apply(aggregate_rscu)\n",
    "data['rscu_aggregate_end'] = data['rscu_end'].apply(aggregate_rscu)\n",
    "data[['primary_id', 'rscu_aggregate_start']].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the standard genetic code\n",
    "codon_table = {\n",
    "    'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "    'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "# Function to split sequence into codons\n",
    "def split_into_codons(sequence):\n",
    "    return [sequence[i:i+3] for i in range(0, len(sequence), 3) if len(sequence[i:i+3]) == 3]\n",
    "\n",
    "# Extract all codons from the nucleotide sequences and calculate RSCU values\n",
    "def calculate_rscu(data, column):\n",
    "    all_codons = []\n",
    "    for seq in data[column]:\n",
    "        all_codons.extend(split_into_codons(seq))\n",
    "\n",
    "    codon_counts = Counter(all_codons)\n",
    "    amino_acid_counts = Counter()\n",
    "    for codon, count in codon_counts.items():\n",
    "        try:\n",
    "            amino_acid = codon_table[codon]\n",
    "            amino_acid_counts[amino_acid] += count\n",
    "        except:\n",
    "            continue\n",
    "        \n",
    "    rscu_values = {codon: (count / amino_acid_counts[codon_table[codon]]) for codon, count in codon_counts.items()}\n",
    "    return rscu_values\n",
    "\n",
    "# Map RSCU values to codons in each sequence\n",
    "def map_rscu_to_codons(seq, rscu_values):\n",
    "    codons = split_into_codons(seq)\n",
    "    return [rscu_values.get(codon, np.nan) for codon in codons]\n",
    "\n",
    "# Extract secondary structure elements from structure string\n",
    "def extract_secondary_structure(structure_string):\n",
    "    return list(structure_string.replace('-', ''))\n",
    "\n",
    "# Align lengths of RSCU values and secondary structure elements by trimming the excess\n",
    "def align_rscu_and_structure(rscu_values, secondary_elements):\n",
    "    min_length = min(len(rscu_values), len(secondary_elements))\n",
    "    return rscu_values[:min_length], secondary_elements[:min_length]\n",
    "\n",
    "# Process each sequence range\n",
    "sequence_ranges = ['start', 'between', 'end']\n",
    "flat_data_list = []\n",
    "\n",
    "for range_ in sequence_ranges:\n",
    "    nucleotide_col = f'nucleotide_sequence_{range_}'\n",
    "    secstru_col = f'secstru_sequence_{range_}'\n",
    "\n",
    "    # Calculate RSCU values\n",
    "    rscu_values = calculate_rscu(data, nucleotide_col)\n",
    "\n",
    "    # Map RSCU values to codons in each sequence\n",
    "    data[f'rscu_values_{range_}'] = data[nucleotide_col].apply(lambda seq: map_rscu_to_codons(seq, rscu_values))\n",
    "\n",
    "    # Extract secondary structure elements\n",
    "    data[f'secondary_structure_elements_{range_}'] = data[secstru_col].apply(extract_secondary_structure)\n",
    "\n",
    "    # Align RSCU values with secondary structure elements\n",
    "    data[f'aligned_rscu_{range_}'], data[f'aligned_secondary_{range_}'] = zip(*data.apply(\n",
    "        lambda row: align_rscu_and_structure(row[f'rscu_values_{range_}'], row[f'secondary_structure_elements_{range_}']), axis=1))\n",
    "\n",
    "    # Flatten the data for correlation analysis\n",
    "    rscu_values_flat = [rscu for sublist in data[f'aligned_rscu_{range_}'] for rscu in sublist]\n",
    "    secondary_elements_flat = [sec for sublist in data[f'aligned_secondary_{range_}'] for sec in sublist]\n",
    "\n",
    "    flat_data = pd.DataFrame({\n",
    "        'RSCU': rscu_values_flat,\n",
    "        'Secondary Structure': secondary_elements_flat,\n",
    "        'Range': range_.capitalize()\n",
    "    })\n",
    "\n",
    "    flat_data_list.append(flat_data)\n",
    "\n",
    "# Combine all flat data into a single DataFrame\n",
    "combined_flat_data = pd.concat(flat_data_list, ignore_index=True)\n",
    "\n",
    "# Plot the combined data\n",
    "plt.figure(figsize=(18, 6))\n",
    "ax = sns.boxplot(x='Range', y='RSCU', hue='Secondary Structure', data=combined_flat_data, palette='viridis', gap=0.2, width=0.95, showfliers=False)\n",
    "    \n",
    "plt.title('Boxplot of RSCU Values by Secondary Structure Element')\n",
    "plt.ylim(0,1.1)\n",
    "plt.legend(loc='upper center', ncol=7)\n",
    "plt.xlabel('Secondary Structure Element')\n",
    "plt.ylabel('RSCU Value')\n",
    "plt.xticks(ticks=range(3), labels=['Start', 'Between', 'End'])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Overlapping KDE Plots\n",
    "# show the density of RSCU values by secondary structure element\n",
    "sns.kdeplot(data=combined_flat_data, x='RSCU', hue='Secondary Structure', fill=True,\n",
    "           palette='viridis', linewidth=0.4)\n",
    "plt.title('KDE Plot of RSCU Values by Secondary Structure Element')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CodonFold",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
