{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import uuid\n",
    "import os\n",
    "\n",
    "from Bio import AlignIO\n",
    "from Bio import SeqIO\n",
    "from Bio import AlignIO\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import scipy.stats as stats\n",
    "\n",
    "from ete3 import Tree, TreeStyle, NodeStyle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate and map the Codon rarity at each position of a multiple sequence alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The calculations are done as in the other script, where the codon rarity score is calculated on the msa and normalized by the number of synonymous codons. Furthermore, gaps in the alignment and then alignment length are taken into account. Unlike the other script, the codon rarity is here not analyzed on a position basis, but for the whole sequence. The codon rarities are summed and devided by the number of codons in the sequence. The tree which is based on the alignment is used to map the codon rarity scores to the leaves of the tree. The phylogenetic tree is colored based on the Codon Rarity Score values."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set data to analyse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "savefig = True # define if figures should be saved\n",
    "\n",
    "# assign the paths to the files need for the analysis\n",
    "alignment_file = \"/Users/dominiquefastus/master_project/NuStru/Example/examples_family/example_fam1_aln.fasta\" \n",
    "nustrudb =  \"/Users/dominiquefastus/master_project/NuStru/Example/examples_family/example_fam1_nustrudb.csv\"\n",
    "tree_file = \"/Users/dominiquefastus/master_project/NuStru/Example/examples_family/example_fam1_tree.tree.rooted\"\n",
    "output_path = \"/Example/examples_family/\"\n",
    "\n",
    "# get a working name from the input file (e.g. protein family name) to use in the output files\n",
    "working_name = os.path.basename(alignment_file)\n",
    "\n",
    "nt_fasta_file = None  # nucleotide sequences in fasta format (optional, so no extra file will be created)\n",
    "\n",
    "# for the temporary created files, like the nucleotide sequences, create a unique job_id\n",
    "job_id = uuid.uuid4()\n",
    "\n",
    "# read the nucleotide structure database as csv\n",
    "nustrudb = pd.read_csv(nustrudb)\n",
    "\n",
    "# read the protein fasta alignment and nucleotide fasta sequences if provided\n",
    "protein_alignment = AlignIO.read(alignment_file, \"fasta\")\n",
    "if nt_fasta_file is not None:\n",
    "    nucleotide_sequences = SeqIO.parse(nt_fasta_file, \"fasta\")\n",
    "else:\n",
    "    # else create a temporary file with the nucleotide sequences corresponding to the protein sequences\n",
    "    for record in protein_alignment:\n",
    "        try:\n",
    "            # get the nucleotide id and sequence from the nustrudb\n",
    "            # and write the id and sequence to a temporary file in the defined output path\n",
    "            nucleotide_id = nustrudb.loc[nustrudb[\"primary_id\"] == record.id][\"nucleotide_id\"].values[0]\n",
    "            nucleotide_sequence = nustrudb.loc[nustrudb[\"primary_id\"] == record.id][\"nucleotide_sequence\"].values[0]\n",
    "            with open(f\"{output_path}/{job_id}_nucleotide_sequences.fasta\", \"a\") as f:\n",
    "                f.write(f\">{nucleotide_id}\\n{nucleotide_sequence}\\n\")\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    # then read the temporary created nucleotide sequences\n",
    "    nucleotide_sequences = SeqIO.parse(f\"{output_path}/{job_id}_nucleotide_sequences.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# note some of the functions are from the msa_codon_structure notebook, they are copied here for the analysis\n",
    "# could be put in a separate file and imported...\n",
    "\n",
    "def fasta_to_array(fasta, align_to=None, codon=False):\n",
    "    \"\"\"Converts a fasta file to a numpy array\"\"\"\n",
    "    # set empty lists to store the sequences and ids\n",
    "    all_seqs = []\n",
    "    all_ids = []\n",
    "    \n",
    "    if codon:\n",
    "        # if the sequences are nucleotide sequences and codons should be used\n",
    "        # then split the sequences into codons for each sequence in the fasta file\n",
    "        for (ind,record) in enumerate(fasta):\n",
    "            all_seqs.append(list(str(record.seq)))\n",
    "            all_seqs[ind] = [''.join(map(str, all_seqs[ind][i:i+3])) for i in range(0, len(all_seqs[ind]), 3)]\n",
    "            \n",
    "    else:         \n",
    "        # else just append the sequences and ids to the lists \n",
    "        # this can be done always for protein sequences  \n",
    "        for record in fasta:\n",
    "            all_seqs.append(list(str(record.seq)))\n",
    "            all_ids.append(record.id)\n",
    "        \n",
    "        # convert the sequences to a numpy array\n",
    "        all_seqs = np.array(all_seqs)\n",
    "\n",
    "    if align_to is not None:\n",
    "        # if an alignment is provided, then insert gaps in the sequences where gaps are in the alignment\n",
    "        # get the gap positions in the alignment and create a numpy array as a mask\n",
    "        gap_indeces = np.where(align_to == '-')\n",
    "        \n",
    "        # loop through the gap positions and insert a gap in the sequences\n",
    "        for gap_index in zip(gap_indeces[0], gap_indeces[1]):\n",
    "            # insert the gap index at the correpsonding position in the sequences\n",
    "            # since we align the codons we use 3 dashes for a gap to show nucleotide triplets\n",
    "            all_seqs[gap_index[0]].insert(gap_index[1], '---')\n",
    "        \n",
    "        # convert the aligned sequences with gaps to a numpy array\n",
    "        all_seqs = np.array(all_seqs)\n",
    "        # deleting stop codons as no protein assigned to them\n",
    "        all_seqs = np.delete(all_seqs, -1, axis=1)\n",
    "    \n",
    "    # reshape the ids to a numpy array according to the number of sequences (one column)\n",
    "    all_ids = np.array(all_ids).reshape(len(all_ids), 1)\n",
    "    \n",
    "    return all_seqs\n",
    "\n",
    "def cub_msa_table(prot_seq_arr=None, cod_seq_arr=None):\n",
    "    \"\"\"Create a codon usage bias table for the multiple sequence alignment\"\"\"\n",
    "    cub_table = {\n",
    "    # '*': {'TAA': None, 'TAG': None, 'TGA': None}, ignoring stop codons\n",
    "    'A': {'GCA': None, 'GCC': None, 'GCG': None, 'GCT': None},\n",
    "    'C': {'TGC': None, 'TGT': None},\n",
    "    'D': {'GAC': None, 'GAT': None},\n",
    "    'E': {'GAA': None, 'GAG': None},\n",
    "    'F': {'TTC': None, 'TTT': None},\n",
    "    'G': {'GGA': None, 'GGC': None, 'GGG': None, 'GGT': None},\n",
    "    'H': {'CAC': None, 'CAT': None},\n",
    "    'I': {'ATA': None, 'ATC': None, 'ATT': None},\n",
    "    'K': {'AAA': None, 'AAG': None},\n",
    "    'L': {'CTA': None, 'CTC': None, 'CTG': None, 'CTT': None, 'TTA': None, 'TTG': None},\n",
    "    'M': {'ATG': None},\n",
    "    'N': {'AAC': None, 'AAT': None},\n",
    "    'P': {'CCA': None, 'CCC': None, 'CCG': None, 'CCT': None},\n",
    "    'Q': {'CAA': None, 'CAG': None},\n",
    "    'R': {'AGA': None, 'AGG': None, 'CGA': None, 'CGC': None, 'CGG': None, 'CGT': None},\n",
    "    'S': {'AGC': None, 'AGT': None, 'TCA': None, 'TCC': None, 'TCG': None, 'TCT': None},\n",
    "    'T': {'ACA': None, 'ACC': None, 'ACG': None, 'ACT': None},\n",
    "    'V': {'GTA': None, 'GTC': None, 'GTG': None, 'GTT': None},\n",
    "    'W': {'TGG': None},\n",
    "    'Y': {'TAC': None, 'TAT': None}}\n",
    "    \n",
    "    for aa in cub_table.keys():\n",
    "        # loop through the amino acids and codons in the cub table and calculate the frequency\n",
    "        # n_AA is the number of occurences of the amino acid in the protein sequence\n",
    "        n_AA = np.count_nonzero(prot_seq_arr == aa)\n",
    "                \n",
    "        # nc_AA is the number of synonyomous codons for the amino acid\n",
    "        nc_AA = len(cub_table[aa].keys())\n",
    "        \n",
    "        for codon in cub_table[aa].keys():\n",
    "            # now for each codon a frequency based on the alignment is calculated\n",
    "            # nc is the number of occurences of the codon in the in the sequences\n",
    "            nc = np.count_nonzero(cod_seq_arr == codon)\n",
    "            \n",
    "            # fc is the frequency of the codon in the alignment\n",
    "            # it is calculated by dividing the number of occurences of the codon by the number of occurences of the amino acid\n",
    "            # then its normalized by the number of synonymous codons\n",
    "            fc =(nc / n_AA) * 1/nc_AA\n",
    "\n",
    "            # round the frequency to 6 decimal places and create an cub table with the frequencies based on the msa\n",
    "            cub_table[aa][codon] = round(fc,6)\n",
    "        \n",
    "    return cub_table\n",
    "\n",
    "def map_rarity(protein_alignment, nustrudb, cu_table):\n",
    "    \"\"\"Map the rarity of the codons to the protein alignment\"\"\"\n",
    "    codon_position_start = 0\n",
    "    \n",
    "    # calculate amino acid occurence in the alignment\n",
    "    # and create a dictionary with the amino acid and its occurence\n",
    "    unique_aa, aa_counts = np.unique(protein_alignment, return_counts=True)\n",
    "    aa_counts = dict(zip(unique_aa, aa_counts))\n",
    "    \n",
    "    # create an empty matrix to store the rarity values\n",
    "    alignment_value_matrix = np.zeros((len(protein_alignment), len(protein_alignment[0])))\n",
    "    seq_name = [seq.id for seq in protein_alignment] # get the sequence names or ids from the alignment\n",
    "    seq_pos = [i for i in range(len(protein_alignment[0]))] # get the sequence positions (with gaps)\n",
    "\n",
    "    # keep track of the position in the alignment for each sequence\n",
    "    # this is needed to adjust the position based on the gaps in the alignment\n",
    "    # so keeps track of the gaps in the alignment\n",
    "    pos_count_dict = {seq_name[i]: 0 for i in range(len(seq_name))}\n",
    "    \n",
    "    # loop over the sequences in the alignment\n",
    "    for position in range(len(protein_alignment[0])):\n",
    "        # loop over the amino acids and sequences in the alignment\n",
    "        for i, (aa, seq) in enumerate(zip(protein_alignment[:,position],seq_name)):\n",
    "            # if the amino acid is a gap, then set the rarity to 0\n",
    "            # else calculate the rarity based on the codon usage table\n",
    "            if aa == '-':\n",
    "                # add gaps to the alignment value matrix\n",
    "                alignment_value_matrix[i, position] = 0 # gaps have no information\n",
    "                pos_count_dict[seq] += 1 # count the gaps for each sequence\n",
    "            else:\n",
    "                # adjust the position in the nucleotide sequence based on the gaps in the alignment\n",
    "                prot_position = pos_count_dict[seq]\n",
    "                position_adj = position - prot_position \n",
    "                \n",
    "                # get the nucleotide sequence from the nustrudb\n",
    "                sequence = nustrudb[nustrudb[\"primary_id\"] == seq][\"nucleotide_sequence\"].values[0]\n",
    "                # set the rarity value based on the codon usage table\n",
    "                # since the codon usage table is based on codons, we need to adjust the position by 3\n",
    "                # the values are then stored in the equivalent position in the alignment value matrix\n",
    "                alignment_value_matrix[i, position] = cu_table[aa][sequence[position_adj*3:position_adj*3+3].upper()]\n",
    "        \n",
    "    # calculate the mean rarity of the sequences in the alignment\n",
    "    # horizontal mean of the rarity values in a row\n",
    "    sequence_mean = {}\n",
    "    for col_mean, name in zip(np.sum(alignment_value_matrix, axis=1), seq_name):\n",
    "        sequence_mean[name] = round((col_mean / len(alignment_value_matrix[0])), 5)\n",
    "    \n",
    "    return sequence_mean\n",
    "\n",
    "def sum_distances_to_root(tree):\n",
    "    \"\"\"Sums the distances of the leaves to the root of the tree\"\"\"\n",
    "    # store the distances in a dictionary with the leaf name as key\n",
    "    leaf_distances = {}\n",
    "    \n",
    "    # loop over the leaves in the tree and calculate the distance to the root\n",
    "    for leaf in tree.iter_leaves():\n",
    "        current, distance_sum = leaf, 0\n",
    "        # while the current node is not the root, sum the distances to the root\n",
    "        while not current.is_root():\n",
    "            distance_sum += current.dist\n",
    "            current = current.up\n",
    "        # round the distance to 5 decimal places and store it in the dictionary\n",
    "        leaf_distances[leaf.name] = round(distance_sum, 5)\n",
    "    return leaf_distances"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the mean Codon Rarity Score for each sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert the fasta files to numpy arrays\n",
    "# split the nucleotide sequences into codons and align them to the protein sequences\n",
    "all_seqs_protein = fasta_to_array(protein_alignment)\n",
    "all_seqs_nt = fasta_to_array(fasta=nucleotide_sequences, align_to=all_seqs_protein, codon=True)\n",
    "\n",
    "# calculate the codon usage table based on the protein and nucleotide sequences\n",
    "# assign the rarity values to the protein alignment in a array\n",
    "cub_msa_table = cub_msa_table(prot_seq_arr=all_seqs_protein, cod_seq_arr=all_seqs_nt)\n",
    "sequence_mean = map_rarity(protein_alignment, nustrudb, cub_msa_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# See relationship between codon rarity and evolution in phylogenetic tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define functions for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_color(value, min_val, max_val):\n",
    "    \"\"\"Determine the color based on the value\"\"\"\n",
    "    # normalize the value between 0 and 1\n",
    "    # creates a color gradient from red to green\n",
    "    norm_val = (value - min_val) / (max_val - min_val)\n",
    "    \n",
    "    # lower values are green, higher values are red\n",
    "    r = int(255 * norm_val)\n",
    "    g = int(255 * (1 - norm_val))\n",
    "    return f'#{r:02x}{g:02x}00'\n",
    "\n",
    "def compute_mean_value(node, leaf_values):\n",
    "    \"\"\"Mean for two or branches\"\"\"\n",
    "    # get the leaf descendants of the node\n",
    "    leaf_descendants = node.get_leaves()\n",
    "    # store the values of the leaf descendants in a list\n",
    "    values = [leaf_values[leaf.name] for leaf in leaf_descendants if leaf.name in leaf_values]\n",
    "    # calculate the mean value\n",
    "    if values:\n",
    "        return sum(values) / len(values)\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regression analysis for the rarity values and the distances to the root of the tree (rarity over divergence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the phylogenetic tree\n",
    "tree = Tree(tree_file)\n",
    "\n",
    "# calculate the distances of the leaves to the root of the tree\n",
    "leaf_distances = sum_distances_to_root(tree)\n",
    "x_values = [leaf_distances[key] for key in sorted(leaf_distances)] # sort the distances for each leaf\n",
    "y_values = [sequence_mean[key] for key in sorted(sequence_mean)] # sort the rarity values for each sequence\n",
    "\n",
    "# convert to numpy arrays\n",
    "x_array = np.array(x_values)\n",
    "y_array = np.array(y_values)\n",
    "\n",
    "# calculate pearson correlation\n",
    "pearson_corr, _ = stats.pearsonr(x_array, y_array)\n",
    "\n",
    "# perform linear regression\n",
    "slope, intercept, r, p, stderr = stats.linregress(x_array, y_array)\n",
    "\n",
    "# determine min and max values for normalization\n",
    "min_val = min(sequence_mean.values())\n",
    "max_val = max(sequence_mean.values())\n",
    "\n",
    "# create a list of colors based on the rarity values\n",
    "colors = [get_color(value, min_val, max_val) for value in y_values]\n",
    "\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.style.use('ggplot')\n",
    "\n",
    "sns.regplot(x=x_values, y=y_values, \n",
    "            scatter_kws={'color': colors, 'alpha': 0.4},  \n",
    "            line_kws={'color': 'navy',})\n",
    "\n",
    "# print and save the pearson correlation\n",
    "print(f\"Pearson Correlation: {round(pearson_corr, 5)}\\n\")\n",
    "with open(f\"{output_path}/{working_name}_divergence_codon_rarity_correlations.txt\", \"w\") as f:\n",
    "    f.write(f\"Pearson Correlation: {round(pearson_corr, 5)}\\n\")\n",
    "\n",
    "plt.xlabel('Branch Lengths from Root (divergence)')\n",
    "plt.ylabel('Codon Rarity Score (sum of all residues)')\n",
    "plt.ylim(0, 1)\n",
    "\n",
    "if savefig:\n",
    "    plt.savefig(f\"{output_path}/{working_name}_divergence_codon_rarity.png\", dpi=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Circular phylogenetic tree colored by codon rarity over the whole sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the leaf values are the rarity values for the sequences\n",
    "leaf_values = sequence_mean\n",
    "\n",
    "# get the min and max values for normalization\n",
    "min_val = min(leaf_values.values())\n",
    "max_val = max(leaf_values.values())\n",
    "\n",
    "# loop over the nodes in the tree and set the style based on the rarity values\n",
    "for node in tree.traverse():\n",
    "    # if the node is a leaf, then set the style based on the rarity value\n",
    "    if node.is_leaf():\n",
    "        # set the color based on the rarity value of the leaf\n",
    "        if node.name in leaf_values:\n",
    "            value = leaf_values[node.name]\n",
    "            color = get_color(value, min_val, max_val)\n",
    "            \n",
    "            lstyle = NodeStyle()\n",
    "            lstyle[\"fgcolor\"] = color\n",
    "            lstyle[\"size\"] = 0\n",
    "            lstyle[\"vt_line_color\"] = color \n",
    "            lstyle[\"hz_line_color\"] = color\n",
    "            lstyle[\"vt_line_width\"] = 0.2\n",
    "            lstyle[\"hz_line_width\"] = 0.2\n",
    "            \n",
    "            node.set_style(lstyle)\n",
    "    else:\n",
    "        # else set the style based on the mean value of the leaf descendants\n",
    "        # this is to color the internal nodes\n",
    "        mean_value = compute_mean_value(node, leaf_values)\n",
    "        if mean_value is not None:\n",
    "            color = get_color(mean_value, min_val, max_val)\n",
    "            \n",
    "        nstyle = NodeStyle()\n",
    "        nstyle[\"size\"] = 0\n",
    "        \n",
    "        nstyle[\"vt_line_color\"] = color\n",
    "        nstyle[\"hz_line_color\"] = color\n",
    "        nstyle[\"vt_line_width\"] = 0.2\n",
    "        nstyle[\"hz_line_width\"] = 0.2\n",
    "        \n",
    "        node.set_style(nstyle)\n",
    "\n",
    "# apply a specific style to the root node (dashed line)\n",
    "rstyle = NodeStyle()\n",
    "rstyle[\"fgcolor\"] = \"black\"\n",
    "rstyle[\"shape\"] = \"square\"\n",
    "rstyle[\"size\"] = 0\n",
    "tree.set_style(rstyle)\n",
    "                    \n",
    "# tree style for circular tree\n",
    "# no branch lengths, no leaf names, no support values\n",
    "ts = TreeStyle()\n",
    "ts.mode = \"c\"\n",
    "ts.show_leaf_name = False\n",
    "ts.show_branch_length = False\n",
    "ts.show_branch_support = False\n",
    "ts.root_opening_factor = 0.1\n",
    "\n",
    "if savefig:\n",
    "    tree.render(f\"{output_path}/{working_name}_tree_codon_rarity.png\", w=1400, h=1600, tree_style=ts, dpi=600)\n",
    "tree.render(\"%%inline\", w=1400, h=1600, tree_style=ts)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "msa_cub",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
