{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import python packages for pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from datetime import date\n",
    "import subprocess\n",
    "import platform\n",
    "import shutil\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Install external softwares"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mmseqs2 already installed:  None\n",
      "mafft already installed:  /opt/homebrew/bin/mafft\n",
      "FastTree already installed:  /opt/homebrew/bin/FastTree\n",
      "FastTree not installed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: brewsci/bio/paml 4.9j is already installed and up-to-date.\n",
      "To reinstall 4.9j, run:\n",
      "  brew reinstall paml\n"
     ]
    }
   ],
   "source": [
    "# Try to install the external programs automatically\n",
    "try:\n",
    "    if shutil.which(\"mmseqs\") is None:\n",
    "        subprocess.run(\"conda install -c conda-forge -c bioconda mmseqs2\", shell=True)\n",
    "        print(\"mmseqs2 not installed\")\n",
    "    elif shutil.which(\"mmseqs\") is not None:\n",
    "        print(\"mmseqs2 already installed: \", shutil.which(\"mmsa\"))\n",
    "except:\n",
    "    print(\"Was not able to install mmseqs2 automatically. Please install manually\")\n",
    "    \n",
    "try:\n",
    "    if \"macOS\" in platform.platform() and shutil.which(\"mafft\") is None:\n",
    "        subprocess.run(\"brew install mafft\", shell=True)\n",
    "        print(\"mafft not installed\")\n",
    "    elif \"Linux\" in platform.platform() and shutil.which(\"mafft\") is None:\n",
    "        print(\"mafft not installed: \", shutil.which(\"mafft\"))\n",
    "    else:\n",
    "        print(\"mafft already installed: \", shutil.which(\"mafft\"))\n",
    "except:\n",
    "    print(\"Was not able to install mafft automatically. Please install manually\")\n",
    "\n",
    "try:\n",
    "    if \"macOS\" in platform.platform() and shutil.which(\"FastTree\") is None:\n",
    "        subprocess.run(\"brew install fasttree\", shell=True)\n",
    "        print(\"FastTree not installed\")\n",
    "    elif \"Linux\" in platform.platform() and shutil.which(\"FastTree\") is None:\n",
    "        subprocess.run(\"sudo apt install fasttree\", shell=True)\n",
    "        print(\"FastTree not installed\")\n",
    "    else:\n",
    "        print(\"FastTree already installed: \", shutil.which(\"FastTree\"))\n",
    "except:\n",
    "    print(\"Was not able to install FastTree automatically. Please install manually\")\n",
    "\n",
    "try:\n",
    "    if \"macOS\" in platform.platform() and shutil.which(\"paml\") is None:\n",
    "        subprocess.run(\"brew install paml\", shell=True)\n",
    "        print(\"FastTree not installed\")\n",
    "    elif \"Linux\" in platform.platform() and shutil.which(\"paml\") is None:\n",
    "        subprocess.run(\"sudo apt install paml\", shell=True)\n",
    "        print(\"FastTree not installed\")\n",
    "    else:\n",
    "        print(\"FastTree already installed: \", shutil.which(\"mmsa\"))\n",
    "except:\n",
    "    print(\"Was not able to install PAML automatically. Please install manually\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup working directory and assign fasta input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to fasta file\n",
    "fasta_file = \"/Users/dominiquefastus/master_project/NuStru/nustruDB/protFAMS/ddl_fam_secstru_reduced.fasta\"\n",
    "\n",
    "job_id = f\"{time.strftime(\"%Y_%m_%d_%H_%M\", time.localtime())}\"\n",
    "# create tree structure for pipeline\n",
    "cwd = os.getcwd()\n",
    "if not os.path.exists(f'{cwd}/nustruTREE/{job_id}/'):\n",
    "    mkdir = os.mkdir(f'{cwd}/nustruTREE/{job_id}/')\n",
    "    mkdir = os.mkdir(f'{cwd}/nustruTREE/{job_id}/MSA/')\n",
    "    mkdir = os.mkdir(f'{cwd}/nustruTREE/{job_id}/TREE/')\n",
    "    \n",
    "shutil.copy(fasta_file, f'{cwd}/nustruTREE/{job_id}/MSA/')\n",
    "\n",
    "# global variables\n",
    "global working_dir \n",
    "global working_name\n",
    "working_dir = f'{cwd}/nustruTREE/{job_id}'\n",
    "working_name = os.path.basename(fasta_file).split(\".\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File copied successfully!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "import time\n",
    "\n",
    "# Path to fasta file\n",
    "fasta_file = \"/Users/dominiquefastus/master_project/NuStru/nustruDB/protFAMS/ddl_fam_secstru_reduced.fasta\"\n",
    "\n",
    "# Generating a unique job ID based on the current time\n",
    "job_id = time.strftime(\"%Y_%m_%d_%H_%M\", time.localtime())\n",
    "\n",
    "# Get the current working directory\n",
    "cwd = os.getcwd()\n",
    "\n",
    "# Create directories for the new job\n",
    "job_path = os.path.join(cwd, 'nustruTREE', job_id)\n",
    "msa_path = os.path.join(job_path, 'MSA')\n",
    "tree_path = os.path.join(job_path, 'TREE')\n",
    "\n",
    "try:\n",
    "    # Ensure all directories exist\n",
    "    os.makedirs(msa_path, exist_ok=True)\n",
    "    os.makedirs(tree_path, exist_ok=True)\n",
    "\n",
    "    # Copy the fasta file to the new directory\n",
    "    shutil.copy(fasta_file, msa_path)\n",
    "    print(\"File copied successfully!\")\n",
    "\n",
    "except FileNotFoundError as fnf_error:\n",
    "    print(f\"Error: {fnf_error}\")\n",
    "    print(\"Please check the path to your fasta file and ensure it exists.\")\n",
    "\n",
    "except PermissionError as perm_error:\n",
    "    print(f\"Permission Error: {perm_error}\")\n",
    "    print(\"Please check your permissions for the directories.\")\n",
    "\n",
    "except Exception as e:\n",
    "    print(f\"An unexpected error occurred: {e}\")\n",
    "\n",
    "# Update global variables\n",
    "global working_dir \n",
    "global working_name\n",
    "working_dir = job_path\n",
    "working_name = os.path.basename(fasta_file).split(\".\")[0]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run the sequence allignment with mmseqs2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db \n",
      "\n",
      "MMseqs Version:       \t15-6f452\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[===\n",
      "Time for merging to ddl_fam_secstru_reduced.fasta.db_h: 0h 0m 0s 8ms\n",
      "Time for merging to ddl_fam_secstru_reduced.fasta.db: 0h 0m 0s 14ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 100ms\n",
      "Create directory /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp\n",
      "linclust /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp --min-seq-id 0.5 -c 0.7 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t15-6f452\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Threads                             \t8\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t2\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.5\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Coverage threshold                  \t0.7\n",
      "Coverage mode                       \t1\n",
      "Max sequence length                 \t65535\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Include identical seq. id.          \tfalse\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "k-mers per sequence                 \t21\n",
      "Spaced k-mers                       \t0\n",
      "Spaced k-mer pattern                \t\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Mask residues                       \t0\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "k-mer length                        \t0\n",
      "Shift hash                          \t67\n",
      "Split memory limit                  \t0\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "\n",
      "Set cluster mode GREEDY MEM.\n",
      "kmermatcher /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 0 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.7 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.5 --kmer-per-seq 21 --spaced-kmer-mode 0 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.7 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 32221 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 32.22K 0s 73ms\n",
      "Sort kmer 0h 0m 0s 18ms\n",
      "Sort by rep. sequence 0h 0m 0s 9ms\n",
      "Time for fill: 0h 0m 0s 9ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 129ms\n",
      "rescorediagonal /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.7 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 32.22K 0s 15ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 0s 33ms\n",
      "clust /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_rescore1 /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 7ms\n",
      "\n",
      "Size of the sequence database: 32221\n",
      "Size of the alignment database: 32221\n",
      "Number of clusters: 6894\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 13ms\n",
      "createsubdb /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/order_redundancy /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "createsubdb /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/order_redundancy /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "filterdb /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_filter1 /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_filter2 --filter-file /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/order_redundancy --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 6.89K 0s 8ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "rescorediagonal /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/input_step_redundancy /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/input_step_redundancy /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_filter2 /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.7 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 6.89K 0s 5ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "align /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/input_step_redundancy /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/input_step_redundancy /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pref_rescore2 /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 2 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.7 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score and coverage\n",
      "Query database size: 6894 type: Aminoacid\n",
      "Target database size: 6894 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 6.89K 2s 529ms\n",
      "Time for merging to aln: 0h 0m 0s 3ms\n",
      "32637 alignments calculated\n",
      "16165 sequence pairs passed the thresholds (0.495297 of overall calculated)\n",
      "2.344793 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 538ms\n",
      "clust /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/input_step_redundancy /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/aln /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 2ms\n",
      "\n",
      "Size of the sequence database: 6894\n",
      "Size of the alignment database: 6894\n",
      "Number of clusters: 3383\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "mergeclusters /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/pre_clust /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/tmp/15027114608538494531/clust --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 6.89K 0s 0ms\n",
      "Clustering step 2\n",
      "[=================================================================] 3.38K 0s 1ms\n",
      "Write merged clustering\n",
      "[=================================================================] 32.22K 0s 4ms\n",
      "Time for merging to cluster: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "result2msa /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster.fasta --msa-format-mode 2 \n",
      "\n",
      "MMseqs Version:           \t15-6f452\n",
      "Substitution matrix       \taa:blosum62.out,nucl:nucleotide.out\n",
      "Gap open cost             \taa:11,nucl:5\n",
      "Gap extension cost        \taa:1,nucl:2\n",
      "Allow deletions           \tfalse\n",
      "Compositional bias        \t1\n",
      "Compositional bias        \t1\n",
      "MSA format mode           \t2\n",
      "Summary prefix            \tcl\n",
      "Skip query                \tfalse\n",
      "Filter MSA                \t0\n",
      "Use filter only at N seqs \t0\n",
      "Maximum seq. id. threshold\t0.9\n",
      "Minimum seq. id.          \t0.0\n",
      "Minimum score per column  \t-20\n",
      "Minimum coverage          \t0\n",
      "Select N most diverse seqs\t1000\n",
      "Preload mode              \t0\n",
      "Threads                   \t8\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Query database size: 32221 type: Aminoacid\n",
      "Target database size: 32221 type: Aminoacid\n",
      "[=================================================================] 3.38K 3s 539ms\n",
      "Time for merging to cluster.fasta: 0h 0m 0s 9ms\n",
      "Time for processing: 0h 0m 3s 571ms\n",
      "createseqfiledb /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster_seq \n",
      "\n",
      "MMseqs Version:\t15-6f452\n",
      "Min sequences\t1\n",
      "Max sequences\t2147483647\n",
      "HH format    \tfalse\n",
      "Preload mode \t0\n",
      "Threads      \t8\n",
      "Compressed   \t0\n",
      "Verbosity    \t3\n",
      "\n",
      "[=================================================================] 3.38K 0s 2ms\n",
      "Time for merging to cluster_seq: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 22ms\n",
      "result2flat /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster_seq /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster_seq.fasta \n",
      "\n",
      "MMseqs Version: \t15-6f452\n",
      "Use fasta header\tfalse\n",
      "Verbosity       \t3\n",
      "\n",
      "Time for processing: 0h 0m 0s 38ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='mmseqs result2flat /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/ddl_fam_secstru_reduced.fasta.db  /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster_seq /Users/dominiquefastus/master_project/NuStru/nustruEVOL/nustruTREE/2024_05_08_20_57/MSA/cluster_seq.fasta', returncode=0)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.run(f\"mmseqs createdb {working_dir}/MSA/{working_name}.fasta {working_dir}/MSA/{working_name}.fasta.db\", shell=True)\n",
    "subprocess.run(f\"mmseqs linclust {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/cluster {working_dir}/MSA/tmp --min-seq-id 0.5 -c 0.7 --cov-mode 1\", shell=True)\n",
    "subprocess.run(f\"mmseqs result2msa {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/cluster {working_dir}/MSA/cluster.fasta --msa-format-mode 2\", shell=True)\n",
    "\n",
    "subprocess.run(f\"mmseqs createseqfiledb {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/cluster {working_dir}/MSA/cluster_seq\", shell=True)\n",
    "subprocess.run(f\"mmseqs result2flat {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/{working_name}.fasta.db  {working_dir}/MSA/cluster_seq {working_dir}/MSA/cluster_seq.fasta\", shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subprocess.run(f\"mmseqs createsubdb {working_dir}/MSA/cluster {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/cluster_repr\", shell=True)\n",
    "subprocess.run(f\"mmseqs convert2fasta {working_dir}/MSA/cluster_repr {working_dir}/MSA/cluster_repr.fasta\", shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subprocess.run(f\"mmseqs align {working_dir}/MSA/cluster_repr {working_dir}/MSA/cluster_repr {working_dir}/MSA/cluster_repr {working_dir}/MSA/cluster_repr_aln\", shell=True)\n",
    "subprocess.run(f\"mmseqs result2msa {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/{working_name}.fasta.db {working_dir}/MSA/cluster_repr_aln {working_dir}/MSA/cluster_repr.msa.fasta --msa-format-mode 2\", shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run the sequence allignment with mafft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subprocess.run(f\"mafft --auto {working_dir}/MSA/{working_name}.fasta > {working_dir}/MSA/{working_name}_aligned.fasta\", shell=True)\n",
    "# subprocess.run(f\"mafft --auto {working_dir}/MSA/cluster_repr.fasta > {working_dir}/MSA/cluster_repr_alligned.fasta\", shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate the phylogenetic tree with FastTree from MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subprocess.run(f\"FastTree {working_dir}/MSA/{working_name}_alligned.fasta > {working_dir}/TREE/cluster_seq.tre\", shell=True)\n",
    "subprocess.run(f\"FastTree {working_dir}/MSA/{working_name}_aligned.fasta > {working_dir}/TREE/{working_name}.tree\", shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CodonFold",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
