{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import ast\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy.stats import ttest_ind\n",
    "import scipy.stats as stats\n",
    "from itertools import combinations\n",
    "from CAI import RSCU"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate and show the RSCU for secondary structure and sequence position"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to calculate RSCU\n",
    "The RSCU is calculate based on Sharp PM, Li WH. An evolutionary perspective on synonymous codon usage in unicellular organisms. Journal of Molecular Evolution. 1986;24(1-2):28â€“38. doi: 10.1007/BF02099948\n",
    "\n",
    "X_ij = observed number of occurrences of the j-th codon for the i-th amino acid, n_i = number of synonymous codons for the i-th amino acid \n",
    "$$ RSCU = \\frac{X_{ij}}{\\frac{1}{n_i} \\sum_{k=1}^{n_j} X_{ij}} $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set data to analyse and split into 3 sequence position (start, between, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "savefig = False # define if figures should be saved\n",
    "\n",
    "# load the data and convert the secondary structure to a list\n",
    "data = pd.read_csv('/Users/dominiquefastus/master_project/NuStru/Example/examples_organism/example_nustru_ecoli.csv', converters={'secondary_structure': ast.literal_eval})   \n",
    "data['secondary_structure'] = data['secondary_structure'].apply(lambda x: list(x.items())) \n",
    "\n",
    "# create new columns for the protein sequence split into 3 parts (start, between, end)\n",
    "data['nucleotide_sequence_start'], data['nucleotide_sequence_between'], data['nucleotide_sequence_end'] = data['nucleotide_sequence'].str[:60], data['nucleotide_sequence'].str[60:-60], data['nucleotide_sequence'].str[-60:]\n",
    "data['protein_sequence_start'], data['protein_sequence_between'], data['protein_sequence_end'] = data['protein_sequence'].str[:20], data['protein_sequence'].str[20:-20], data['protein_sequence'].str[-20:]\n",
    "# due to the splitting to 20 positions start and end, \n",
    "# there will be empty rows for short sequeneces, so we delete them (not optimal)\n",
    "data = data[data['nucleotide_sequence_between']!= '']\n",
    "\n",
    "# create new columns for the secondary structure split into 3 parts (start, between, end)\n",
    "# convert the secondary structure dictionary to a string (whole sequence and ranges)\n",
    "data['secstru_sequence_start'], data['secstru_sequence_between'], data['secstru_sequence_end'] = data['secondary_structure'].apply(lambda x: x[:20]), data['secondary_structure'].apply(lambda x: x[20:-20]), data['secondary_structure'].apply(lambda x: x[-20:])\n",
    "data['secstru_sequence_start'], data['secstru_sequence_between'], data['secstru_sequence_end'] = data['secstru_sequence_start'].apply(lambda x: \"\".join(dict(x).values())), data['secstru_sequence_between'].apply(lambda x: \"\".join(dict(x).values())), data['secstru_sequence_end'].apply(lambda x: \"\".join(dict(x).values()))\n",
    "data['secondary_structure'] = data['secondary_structure'].apply(lambda x: \"\".join(dict(x).values()))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define functions and codon table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the standard genetic code with stop codons\n",
    "codon_table = {\n",
    "    'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "    'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "def split_into_codons(sequence):\n",
    "    \"\"\"Split the sequence into codons\"\"\"\n",
    "    # codons are a subset of 3 nucleotide, so we check again if the sequence is divisible by 3\n",
    "    # due to previous filtering, this should be always the case\n",
    "    return [sequence[i:i+3] for i in range(0, len(sequence), 3) if len(sequence[i:i+3]) == 3]\n",
    "\n",
    "def map_rscu_to_codons(seq, rscu_values):\n",
    "    \"\"\"Map RSCU values to codons in each sequence\"\"\"\n",
    "    # get a list of the codons of the nucleotide sequence (same order)\n",
    "    codons = split_into_codons(seq)\n",
    "\n",
    "    # return the calculated rscu values of each codon in the sequence\n",
    "    mapped_rscu_value = [rscu_values.get(codon, np.nan) for codon in codons]\n",
    "\n",
    "    # most RSCU calculations ignore the stop codons and since it is of no interest for the analysis, it is deleted\n",
    "    if mapped_rscu_value[-1] in ['TAA','TAG','TGA']:\n",
    "        mapped_rscu_value.pop()\n",
    "\n",
    "    return mapped_rscu_value\n",
    "\n",
    "# this was an initial function with the implemented rscu calculation\n",
    "# to be more consistend with previous publications, it was choosen to use the CAI package\n",
    "def calculate_rscu(seq_given=None, data=None, column=None):\n",
    "    \"\"\"Calculate RSCU values for the sequence (gene)\"\"\"\n",
    "    \n",
    "    # setup a list with the codons of the sequence\n",
    "    # depending if only one or multiple sequences are given\n",
    "    all_codons = []\n",
    "    if isinstance(data, pd.DataFrame):\n",
    "        for seq_row in data[column]:\n",
    "            all_codons.extend(split_into_codons(seq_row))\n",
    "    else:\n",
    "        all_codons = split_into_codons(seq_given)\n",
    "\n",
    "    # calculate the xth codon for the ith amino acid\n",
    "    # setup a counter for both and get the absolute count from the codon list\n",
    "    codon_counts = Counter(all_codons)\n",
    "    amino_acid_counts = Counter()\n",
    "\n",
    "    # Count the codon belonging to the same amino acid (number of amino acids)\n",
    "    for codon, count in codon_counts.items():\n",
    "        if codon in codon_table:\n",
    "            # if the codon is in the codon table then get the belonging amino acid\n",
    "            # add the count to the amino acid to get a total count\n",
    "            # this is needed to calculate the expected codon usage\n",
    "            amino_acid = codon_table[codon]\n",
    "            amino_acid_counts[amino_acid] += count\n",
    "\n",
    "    # now calculate the rscu values for each codon and store them in a dictionaryy\n",
    "    rscu_values = {}\n",
    "    for codon, count in codon_counts.items():\n",
    "        if codon in codon_table:\n",
    "            # get the number of synonymous codons for a codon, belonging to one amino acid\n",
    "            amino_acid = codon_table[codon]\n",
    "            synonymous_codons = [c for c, a in codon_table.items() if a == amino_acid]\n",
    "\n",
    "            # the expected count or usage is the usage if all codons would be uniform and equally used\n",
    "            # so we calculate it by the total counts of the codons, devided by the number of synonymous codons\n",
    "            expected_count = amino_acid_counts[amino_acid] / len(synonymous_codons)\n",
    "\n",
    "            # we can then calculate the rscu with the actual count of the codon (observed codon usage),\n",
    "            # divided by the expected count (expted codon usage)\n",
    "            rscu_values[codon] = (count / expected_count)\n",
    "\n",
    "    # create a list of the rscu based on the codon order in the sequence\n",
    "    rscu_results = map_rscu_to_codons(seq, rscu_values)\n",
    "\n",
    "    if column == 'nucleotide_sequence_end':\n",
    "        # most RSCU calculations ignore the stop codons and since it is of no interest for the analysis, it is deleted\n",
    "        rscu_results.pop() \n",
    "        \n",
    "    return rscu_results\n",
    "\n",
    "def rscu_method_calculation(seq_given=None, data=None, column=None):\n",
    "    \"\"\"Calculates the RSCU values based on the CAI package\"\"\"\n",
    "    # setup a list of sequences to calculate the codons from\n",
    "    if isinstance(data, pd.DataFrame):\n",
    "        seq = data[column].values.tolist()\n",
    "    else:\n",
    "        seq = [str(seq_given)]\n",
    "\n",
    "    # there are multiple ways to calculate the rscu for codons based on literature\n",
    "    # while rscu values for one sequence can indicate sequence (protein specific codon usage)\n",
    "    # drawing the rscu from a set can give rise to species or overall codon usage\n",
    "    if len(seq) == 1:\n",
    "        # calculate the rscu for codons based on the sequence\n",
    "        rscu_values = RSCU(seq)\n",
    "\n",
    "        # create a list of the rscu based on the codon order in the sequence\n",
    "        rscu_results = map_rscu_to_codons(seq_given, rscu_values)\n",
    "        \n",
    "        if column == 'nucleotide_sequence_end':\n",
    "            # most RSCU calculations ignore the stop codons and since it is of no interest for the analysis, it is deleted\n",
    "            rscu_results.pop() \n",
    "        \n",
    "    elif len(seq) >= 1:\n",
    "        # calculate the rscu for codons based on multiple sequences\n",
    "        rscu_values = RSCU(seq)\n",
    "        rscu_results = rscu_values\n",
    "    else:\n",
    "        print(\"Can't calculate rscu as no sequences provivded!\")\n",
    "\n",
    "    return rscu_results\n",
    "\n",
    "def extract_secondary_structure(structure_string):\n",
    "    \"\"\"Extract secondary structure elements from structure string\"\"\"\n",
    "    # convert string to list, a function is needed as it is applied to each row individually\n",
    "    return list(structure_string) \n",
    "\n",
    "def align_rscu_and_structure(rscu_values, secondary_elements):\n",
    "    \"\"\" Align lengths of RSCU values and secondary structure elements by trimming the excess\"\"\"\n",
    "    # sometimes the sequence don't align equally\n",
    "    # this can be done nicer and quicker (so it's probably redundant)\n",
    "    min_length = min(len(rscu_values), len(secondary_elements))\n",
    "    return rscu_values[:min_length], secondary_elements[:min_length]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate the RSCU for each sequence and extract the information for the range and structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# process each sequence range\n",
    "# create a data list to store the ranges with sequences and rscu values\n",
    "# 1seq indicates that the rscu is calculated for one sequence\n",
    "sequence_ranges = ['start', 'between', 'end']\n",
    "flat_data_list_1seq = []\n",
    "flat_data_list_rare_1seq = [] # we can also extract rare or no frequently used codons and store them here\n",
    "\n",
    "# map RSCU values to codons in each sequence\n",
    "data[f'rscu_values_sequence'] = data['nucleotide_sequence'].apply(lambda seq: rscu_method_calculation(seq_given=seq))\n",
    "\n",
    "# loop over the ranges and map the rscu values to the secondary structures\n",
    "for range_ in sequence_ranges:\n",
    "\n",
    "    # Extract secondary structure elements for the individual ranges\n",
    "    # transform elements to a list\n",
    "    secstru_col = f'secstru_sequence_{range_}'\n",
    "    data[f'secondary_structure_elements_{range_}'] = data[secstru_col].apply(extract_secondary_structure)\n",
    "\n",
    "    # Align RSCU values with secondary structure elements\n",
    "    # do it for the individual ranges\n",
    "    if range_ == 'start':\n",
    "        data[f'aligned_rscu_{range_}'], data[f'aligned_secondary_{range_}'] = zip(*data.apply(\n",
    "            lambda row: align_rscu_and_structure(row[f'rscu_values_sequence'][:20], row[f'secondary_structure_elements_{range_}']), axis=1))\n",
    "    elif range_ == 'between':\n",
    "        data[f'aligned_rscu_{range_}'], data[f'aligned_secondary_{range_}'] = zip(*data.apply(\n",
    "            lambda row: align_rscu_and_structure(row[f'rscu_values_sequence'][20:-20], row[f'secondary_structure_elements_{range_}']), axis=1))\n",
    "    elif range_ == 'end':\n",
    "        data[f'aligned_rscu_{range_}'], data[f'aligned_secondary_{range_}'] = zip(*data.apply(\n",
    "            lambda row: align_rscu_and_structure(row[f'rscu_values_sequence'][-20:], row[f'secondary_structure_elements_{range_}']), axis=1))\n",
    "\n",
    "    # flatten the data for correlation analysis by adding it to a list and transform it again to a dataframe\n",
    "    rscu_values_flat_1seq = [rscu for sublist in data[f'aligned_rscu_{range_}'] for rscu in sublist]\n",
    "    secondary_elements_flat_1seq = [sec for sublist in data[f'secondary_structure_elements_{range_}'] for sec in sublist]\n",
    "\n",
    "    flat_data_1seq = pd.DataFrame({\n",
    "        'RSCU': rscu_values_flat_1seq,\n",
    "        'Secondary Structure': secondary_elements_flat_1seq,\n",
    "        'Range': range_.capitalize()\n",
    "    })\n",
    "\n",
    "    # append the flat data so we get a complete \"transpose\" dataframe\n",
    "    flat_data_list_1seq.append(flat_data_1seq)\n",
    "    \n",
    "    # filter RSCU values less than 0.6, which are considered to be underrepresented\n",
    "    flat_data_rare_1seq = flat_data_1seq[flat_data_1seq['RSCU'] < 0.6]\n",
    "    flat_data_list_rare_1seq.append(flat_data_rare_1seq)\n",
    "\n",
    "# combine all flat data into a single DataFrame\n",
    "# this make it possible to plot the data\n",
    "combined_flat_data_1seq = pd.concat(flat_data_list_1seq, ignore_index=True)\n",
    "combined_flat_data_rare_1seq = pd.concat(flat_data_list_rare_1seq, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate RSCU based on a set of sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process each sequence range\n",
    "flat_data_list_nseq = []\n",
    "flat_data_list_rare_nseq = []\n",
    "\n",
    "# loop over the ranges and map the rscu values to the secondary structures\n",
    "for range_ in sequence_ranges:\n",
    "    nucleotide_col = f'nucleotide_sequence_{range_}'\n",
    "    secstru_col = f'secstru_sequence_{range_}'\n",
    "\n",
    "    # Calculate RSCU values\n",
    "    rscu_values_nseq = rscu_method_calculation(data=data, column=nucleotide_col)\n",
    "\n",
    "    # Map RSCU values to codons in each sequence\n",
    "    data[f'nseq_rscu_values_{range_}'] = data[nucleotide_col].apply(lambda seq: map_rscu_to_codons(seq, rscu_values_nseq))\n",
    "\n",
    "    # Align RSCU values with secondary structure elements\n",
    "    data[f'nseq_aligned_rscu_{range_}'], data[f'nseq_aligned_secondary_{range_}'] = zip(*data.apply(\n",
    "        lambda row: align_rscu_and_structure(row[f'nseq_rscu_values_{range_}'], row[f'secondary_structure_elements_{range_}']), axis=1))\n",
    "\n",
    "    # Flatten the data for correlation analysis\n",
    "    rscu_values_flat_nseq = [rscu for sublist in data[f'nseq_aligned_rscu_{range_}'] for rscu in sublist]\n",
    "    secondary_elements_flat_nseq = [sec for sublist in data[f'nseq_aligned_secondary_{range_}'] for sec in sublist]\n",
    "\n",
    "    flat_data_nseq = pd.DataFrame({\n",
    "        'RSCU': rscu_values_flat_nseq,\n",
    "        'Secondary Structure': secondary_elements_flat_nseq,\n",
    "        'Range': range_.capitalize()\n",
    "    })\n",
    "\n",
    "    # append the flat data so we get a complete \"transpose\" dataframe\n",
    "    flat_data_list_nseq.append(flat_data_nseq)\n",
    "    \n",
    "    # filter RSCU values less than 0.6, which are considered to be underrepresented\n",
    "    flat_data_rare_nseq = flat_data_nseq[flat_data_nseq['RSCU'] < 0.6]\n",
    "    flat_data_list_rare_nseq.append(flat_data_rare_nseq)\n",
    "\n",
    "# combine all flat data into a single DataFrame\n",
    "# this make it possible to plot the data\n",
    "combined_flat_data_nseq = pd.concat(flat_data_list_nseq, ignore_index=True)\n",
    "combined_flat_data_rare_nseq = pd.concat(flat_data_list_rare_nseq, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualize the RSCU values for the secondary structure in the ranges, show the rare RSCU destribution seperately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# please choose if the rscu calculated from one sequence or a set of sequence should be plotted and analysed\n",
    "combined_flat_data = combined_flat_data_nseq\n",
    "combined_flat_data_rare = combined_flat_data_rare_nseq\n",
    "plot_and_data_name = \"nseq\"\n",
    "\n",
    "# set if the rscu is calculated from one sequence or a set of sequences\n",
    "rscu_from_one_seq = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the combined data for the overall rscu destribuation for each secondary structure\n",
    "# same groups of secondary structure get same color\n",
    "custom_palette = {'G': '#FFA500', 'I': '#FF8C00', 'H': '#e0771b', 'S': '#1E90FF', \n",
    "                  'T': '#00BFFF', '-': '#4682B4', 'E': '#32CD32', 'B': '#3CB371'} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(18,6))\n",
    "sns.boxplot(x='Range', y='RSCU', hue='Secondary Structure', data=combined_flat_data, \n",
    "                  palette=custom_palette, width=0.95, gap=0.2, showfliers=False)\n",
    "\n",
    "plt.legend(loc='upper center', ncol=8)\n",
    "plt.ylabel('RSCU Value')\n",
    "plt.ylim(0, 3.5)\n",
    "plt.xlabel('Secondary Structure Element')\n",
    "plt.xticks(ticks=range(3), labels=['Start', 'Between', 'End'])\n",
    "\n",
    "if savefig:\n",
    "    plt.savefig(f'{plot_and_data_name}_RSCU_vs_secondary_structure.png', dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(18,4))\n",
    "ax = sns.boxplot(x='Range', y='RSCU', hue='Secondary Structure', data=combined_flat_data_rare, \n",
    "                  palette=custom_palette, width=0.95, gap=0.2, showfliers=False)\n",
    "\n",
    "ax.legend_.remove()\n",
    "plt.ylabel('RSCU Value')\n",
    "plt.xlabel('Secondary Structure Element')\n",
    "plt.xticks(ticks=range(3), labels=['Start', 'Between', 'End'])\n",
    "\n",
    "if savefig:\n",
    "    plt.savefig(f'{plot_and_data_name}_RSCU_vs_secondary_structure_rare.png', dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the overview statistics of the barplots as spreadsheets\n",
    "grouped_data = combined_flat_data.groupby(['Range', 'Secondary Structure'])\n",
    "grouped_data.describe().to_excel(f'{plot_and_data_name}RSCU_values_by_secondary_structure_element.xlsx')\n",
    "\n",
    "grouped_rare_data = combined_flat_data_rare.groupby(['Range', 'Secondary Structure'])\n",
    "grouped_rare_data.describe().to_excel(f'{plot_and_data_name}_Rare_RSCU_values_by_secondary_structure_element.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform a t_test and one_way anova to see if the RSCU values are significant between structure within a range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize empty lists to store results for all ranges\n",
    "all_p_values = []\n",
    "all_comparisons = []\n",
    "all_ranges = []\n",
    "\n",
    "# the ranges from previous assigned again for reference\n",
    "ranges = [\"Start\", \"Between\", \"End\"]\n",
    "\n",
    "for range_name in ranges:\n",
    "    # get the data for the current range\n",
    "    # get the secondary structures (were also defined earlier)\n",
    "    combined_flat_data_range = combined_flat_data[combined_flat_data[\"Range\"] == range_name]\n",
    "    groups = combined_flat_data_range[\"Secondary Structure\"].unique()\n",
    "\n",
    "    # calculate p-values for each pair of groups within the current range by conducting a t_test\n",
    "    for group1, group2 in combinations(groups, 2):\n",
    "        # loop over all combinations within the ranges and retrieve the RSCU of the group\n",
    "        group1_data = combined_flat_data_range[combined_flat_data_range[\"Secondary Structure\"] == group1][\"RSCU\"]\n",
    "        group2_data = combined_flat_data_range[combined_flat_data_range[\"Secondary Structure\"] == group2][\"RSCU\"]\n",
    "\n",
    "        # apply the ttest and get the t_stat and p_value\n",
    "        t_stat, p_value = ttest_ind(group1_data, group2_data)\n",
    "        all_p_values.append(p_value) # add the p value\n",
    "        all_comparisons.append((group1, group2)) # add the comparison of the group\n",
    "        all_ranges.append(range_name) # add the range\n",
    "\n",
    "# create a DataFrame for the results\n",
    "results = pd.DataFrame(all_comparisons, columns=[\"Group 1\", \"Group 2\"])\n",
    "results[\"Range\"] = all_ranges\n",
    "results[\"P Value\"] = all_p_values\n",
    "\n",
    "# apply Bonferroni correction\n",
    "# since we do multi comparisons analysis, the p-value is adjusted for type 1 errors\n",
    "alpha = 0.05\n",
    "m = len(all_p_values)  # number of comparisons\n",
    "adjusted_alpha = alpha / m\n",
    "\n",
    "# check if the rscu between secondary structures are significant\n",
    "results[\"Significant\"] = results[\"P Value\"] < adjusted_alpha\n",
    "results.to_csv(f'{plot_and_data_name}_t_test_of_secstru_rscu_within_range.csv')\n",
    "\n",
    "print(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for range_name in ranges:\n",
    "    # group the secondary structures and save the rscu values to a list\n",
    "    combined_flat_data_range = combined_flat_data[combined_flat_data[\"Range\"] == range_name]\n",
    "    g = combined_flat_data_range.groupby(\"Secondary Structure\")[\"RSCU\"].apply(list)\n",
    "\n",
    "    # calculate a oneway anova for the rscu values of the secondary structure\n",
    "    f_stat, p_value = stats.f_oneway(*g)\n",
    "    \n",
    "    print(f\"Anova result for {range_name}\")\n",
    "    print(f\"f_stat: {f_stat}, {p_value:e}\")\n",
    "    print(f\"Significant: {p_value < adjusted_alpha}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare and visualize the RSCU between the secondary structure of the ranges "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# aggregate the rscu values by secondary structure element and range (take the mean)\n",
    "# also aggregat the rare rscu values (not plotted here, but were visuealized in the results)\n",
    "aggregated_data = combined_flat_data.groupby(['Range', 'Secondary Structure'])['RSCU'].mean().unstack()\n",
    "aggregated_data_rare = combined_flat_data_rare.groupby(['Range', 'Secondary Structure'])['RSCU'].mean().unstack()\n",
    "\n",
    "# fill any missing values with 0 (or another appropriate value)\n",
    "aggregated_data = aggregated_data.fillna(0)\n",
    "aggregated_data_rare = aggregated_data_rare.fillna(0)\n",
    "\n",
    "# set the labels (the secondary structures from earlier)\n",
    "# and get the ranges also from earlier\n",
    "labels = aggregated_data.columns.tolist()\n",
    "ranges = aggregated_data.index.tolist()\n",
    "\n",
    "# number of variables to span the radarplot\n",
    "num_vars = len(labels)\n",
    "\n",
    "# compute angle for each axis from the number of labels to get a circle\n",
    "angles = np.linspace(0, 2 * np.pi, num_vars, endpoint=False).tolist()\n",
    "\n",
    "# repeat the first value to close the circular graph\n",
    "angles += angles[:1]\n",
    "\n",
    "# initialize the radar plot\n",
    "# create subplots of each range\n",
    "fig, ax = plt.subplots(figsize=(8, 8), subplot_kw=dict(polar=True))\n",
    "\n",
    "# draw one axe per variable and add labels\n",
    "ax.set_theta_offset(np.pi / 2)\n",
    "ax.set_theta_direction(-1)\n",
    "plt.xticks(angles[:-1], labels) # the x_ticks are the labels of the circle\n",
    "\n",
    "# create a color palette based on the number of ranges\n",
    "colors = plt.cm.viridis(np.linspace(0, 1, len(ranges)))\n",
    "\n",
    "# plot each range's data for the aggregated data\n",
    "for range_, color in zip(ranges, colors):\n",
    "    values = aggregated_data.loc[range_].tolist() # get the values\n",
    "    values += values[:1] # repeat the first values to connect the line\n",
    "\n",
    "    # plot the line plot and fill the area\n",
    "    ax.plot(angles, values, linewidth=2, linestyle='solid', label=range_, color=color)\n",
    "    ax.fill(angles, values, color=color, alpha=0.2)\n",
    "\n",
    "plt.legend(loc='upper right', bbox_to_anchor=(1.1, 1.1))\n",
    "\n",
    "if savefig:\n",
    "    plt.savefig(f'{plot_and_data_name}_ranges_rscu_radaplot.png', dpi=600)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the statistics of the radarplot as a spreadsheet\n",
    "aggregated_data.groupby(\"Range\").describe().to_excel(f'{plot_and_data_name}_RSCU_values_by_range.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform one_way anova to see if the RSCU values are significant between ranges and secondary structure elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare a dictionary to store ANOVA results\n",
    "results_anova = {}\n",
    "\n",
    "for structure in labels:\n",
    "    # for each secondary structure, get the data\n",
    "    structure_data = combined_flat_data[['Range', 'Secondary Structure', 'RSCU']]\n",
    "    structure_data = structure_data[structure_data['Secondary Structure'] == structure]\n",
    "    \n",
    "    # prepare data for ANOVA\n",
    "    data_groups = [structure_data[structure_data['Range'] == range_]['RSCU'] for range_ in ranges]\n",
    "    \n",
    "    # perform ANOVA\n",
    "    f_stat, p_val = stats.f_oneway(*data_groups)\n",
    "    \n",
    "    results_anova[structure] = (f_stat, p_val)\n",
    "\n",
    "for key, value in results_anova.items():\n",
    "    print(f'{key}: F-statistic = {value[0]}, p-value = {value[1]}')\n",
    "    print(f\"Significant: {p_value < 0.05}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate the RSCU along the structure position for Coils, Sheets and Helices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Here the secondary structure types are aggregated as: Coil (T,S,-), Sheet (E,B) and Helix (H,I,G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the structure was only extracted from ranges and not in total, so it's done here\n",
    "data[f'secondary_structure_elements'] = data['secondary_structure'].apply(extract_secondary_structure)\n",
    "\n",
    "if rscu_from_one_seq:\n",
    "    # align the rscu values again with the secondary structure elements\n",
    "    data[f'aligned_rscu'], data[f'aligned_secondary'] = zip(*data.apply(\n",
    "        lambda row: align_rscu_and_structure(row[f'rscu_values_sequence'], row[f'secondary_structure_elements']), axis=1))\n",
    "    \n",
    "else:\n",
    "    # concatenate the rscu values from the start, between and end\n",
    "    data[f'nseq_aligned_rscu_total'] = data.apply(lambda row: row[f'nseq_rscu_values_start'] + row[f'nseq_rscu_values_between'] + row[f'nseq_rscu_values_end'], axis=1)\n",
    "    \n",
    "    # align the rscu values again with the secondary structure elements\n",
    "    data[f'aligned_rscu'], data[f'aligned_secondary'] = zip(*data.apply(\n",
    "        lambda row: align_rscu_and_structure(row[f'nseq_aligned_rscu_total'], row[f'secondary_structure_elements']), axis=1))\n",
    "\n",
    "# aggregate the secondary structures from dssp to helix, coil and sheet\n",
    "results = []\n",
    "secstru_aggregation = {\"H\": \"Helix\", \"I\": \"Helix\", \"G\": \"Helix\", \"E\": \"Sheet\", \"B\": \"Sheet\", \"T\": \"Coil\", \"S\": \"Coil\", \"-\": \"Coil\"}\n",
    "\n",
    "# get all segments of adjacent secondary structures\n",
    "for _, row in data.iterrows():\n",
    "    current_segment = []\n",
    "    current_secstru = None\n",
    "    position = 1 # set a position counter\n",
    "\n",
    "    for rscu, secstru in zip(row['aligned_rscu'], row['aligned_secondary']):\n",
    "        secstru = secstru_aggregation[secstru] # assign the overall secondary structure\n",
    "\n",
    "        # if the secondary structure is the same as the previous, then add to the segment\n",
    "        if secstru == current_secstru: \n",
    "            # add to the segment\n",
    "            current_segment.append((secstru, position, rscu))\n",
    "        # when the current segment is over, so a new secondary structure motif is present \n",
    "        # then check the length of the segment\n",
    "        else:\n",
    "            if current_segment:\n",
    "                # the segments should not be long than 70 positions (but can be changed otherwise)\n",
    "                if not ((current_secstru == \"Helix\" and len(current_segment) >= 70) or\n",
    "                        (current_secstru == \"Sheet\" and len(current_segment) >= 70) or\n",
    "                        current_secstru == \"Coil\" and len(current_segment) >= 70):\n",
    "\n",
    "                    # add the segment to the results\n",
    "                    results.extend(current_segment)\n",
    "\n",
    "            # set the new segment with the structure element\n",
    "            current_segment = [(secstru, 1, rscu)]\n",
    "            current_secstru = secstru\n",
    "            position = 1  # reset position for new segment\n",
    "        position += 1 # increase the position within a segment\n",
    "\n",
    "# create a dataframe to plot the data\n",
    "df_rscu_position = pd.DataFrame(results, columns=['Secondary Structure', 'Position', 'RSCU'])\n",
    "\n",
    "\n",
    "# we assume that segments are similar long, so we can apply a rolling mean to smooth the data\n",
    "df_rscu_position['Smoothed_RSCU'] = df_rscu_position['RSCU'].rolling(window=3, min_periods=1, center=True).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a line plot with the standard deviation\n",
    "sns.lineplot(x=\"Position\", y=\"Smoothed_RSCU\", hue=\"Secondary Structure\", data=df_rscu_position)\n",
    "plt.ylim(1,2)\n",
    "if savefig: \n",
    "    plt.savefig('RSCU_by_position.png')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CodonFold",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
